### PR TITLE
Use Dyninst for x86_64 classification of types

### PR DESCRIPTION
**Pull request by hainest**
There are three big changes here:

1. Use Dyninst's type information to classify language types into ABI types
2. Expand the register allocator to handle more cases
3. Use a Python script to generate tests and test functions

There are a _lot_ of TODOs hanging around because I kept running into corner cases that I don't have a good answer to right now. We'll need to eventually figure them out, but some of them are the same problems that the LLVM folks haven't implemented yet, either.

### CODE REVIEW

File: source/parser/x86_64/allocators.hpp [hunk 1]
```diff
@@ -53,35 +54,59 @@ namespace smeagle::x86_64 {
...
+        }
+        return reg.value();
+      }
 
-      // If the class is memory, pass the argument on the stack
-      if (regClass == RegisterClass::NO_CLASS) regString = "";
-      if (regClass == RegisterClass::SSE) regString = this->getNextSseRegister();
-      if (regClass == RegisterClass::INTEGER) regString = this->getNextIntRegister();
-      if (regClass == RegisterClass::MEMORY) regString = std::nullopt;
+      if (lo == RegisterClass::SSE) {
```
Comment (vsoch) [inline source/parser/x86_64/allocators.hpp:76]: Is there any reason to not have these as their own functions (like before?)

File: source/parser/x86_64/allocators.hpp [hunk 2]
```diff
@@ -53,35 +54,59 @@ namespace smeagle::x86_64 {
...
+
+      if (lo == RegisterClass::X87) {
+        // goes on the stack
+        return fallocator.nextFramebaseFromType(paramType);
+      }
 
-      // If we don't have a value, we need a framebase
-      if (!regString.has_value()) {
-        regString = fallocator.nextFramebaseFromType(paramType);
+      if (lo == RegisterClass::COMPLEX_X87) {
```
Comment (vsoch) [inline source/parser/x86_64/allocators.hpp:103]: Could this statement be put as an OR on line 98 (or do you anticipate more changes?)

File: source/parser/x86_64/classifiers.hpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+#include <utility>
+
+#include "Type.h"
+#include "register_class.hpp"
+
+namespace smeagle::x86_64 {
+
+  struct Class {
+    RegisterClass lo, hi;
```
Comment (vsoch) [inline source/parser/x86_64/classifiers.hpp:16]: Could there ever be cases in the future of having more than two registers that would warrant this approach a not great idea?

File: source/parser/x86_64/classifiers.hpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+       * 	3)	A recursive conversion of application type X, at location “(base location)”
+       * 	(the base location wrapped in parentheses) and with direction ‘Import’.
+       *
+       * 	For example, an application type int* at base location %rax and direction ‘Import’
+       * would convert to: abi_typelocation(..., Import, Pointer64, “%rax”). abi_typelocation(...,
+       * Export, Integer32, “(%rax)”). abi_typelocation(..., Import, Integer32, “(%rax)”).
+       *
+       */
+      // TODO Should we integrate the directionality calculation here?
+      return {RegisterClass::INTEGER, RegisterClass::NO_CLASS, ptr_cnt, "Pointer64"};
```
Comment (vsoch) [inline source/parser/x86_64/classifiers.hpp:44]: Why is this the entire block to return this hard coded thing? (is the logic just missing where the text is now?)

File: source/parser/x86_64/classifiers.hpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                "CplxFloat" + std::to_string(size / 2)};
+      }
+      if (size <= 64) {
+        // 32- or 64-bit floats
+        return {RegisterClass::SSE, RegisterClass::SSEUP, ptr_cnt, "Float" + std::to_string(size)};
+      }
+      if (size == 128) {
+        // x87 `long double` OR __m128[d]
+        // TODO: How do we differntiate the vector type here? Dyninst should help us
+        return {RegisterClass::X87, RegisterClass::X87UP, ptr_cnt, "Float128"};
```
Comment (vsoch) [inline source/parser/x86_64/classifiers.hpp:88]: Is there somewhere we are writing down this standard / dictionary of types we are going to use?

File: source/parser/x86_64/x86_64.cpp [hunk 6]
```diff
@@ -67,85 +68,37 @@ namespace smeagle::x86_64 {
...
-    paramType = remove_typedef(paramType);
+  Class getRegisterClassFromType(st::Type *paramType) {
+    auto [base_type, ptr_cnt] = unwrap_underlying_type(paramType);
 
-    // If it's a pointer, remove it
-    if (is_indirect(paramType->getDataClass())) {
-      paramType = deref(paramType);
-      paramType = remove_typedef(paramType);
+    if (auto *t = base_type->getScalarType()) {
+      return classify(t, ptr_cnt);
```
Comment (vsoch) [inline source/parser/x86_64/x86_64.cpp:75]: I like this a lot better :)

File: source/parser/x86_64/classifiers.hpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+  struct Class {
+    RegisterClass lo, hi;
+    int pointer_indirections;
+    std::string name;
+  };
+
+  namespace st = Dyninst::SymtabAPI;
+
+  inline Class classify(st::typeScalar *t, int ptr_cnt) {
```
Comment (vsoch) [inline source/parser/x86_64/classifiers.hpp:23]: Hmm maybe call this a RegisterClass to not confuse with what we normally think of as a class?

Comment (vsoch) [reply source/parser/x86_64/classifiers.hpp:23]: oh wait a Class has a RegisterClass... ohno.  What about if we just reverse it? As long as the Class is internal it won't be as weird to instantiate something called just Class?

```cpp
struct RegisterClass {
    Class lo, hi;
}
```
or even just

```cpp
struct RegisterClass {
    ClassName lo, hi;
}
```
or 

```cpp
struct RegisterClasses {
    RegisterClass lo, hi;
}
```
Maybe it's not a big deal- what are your thoughts?

File: source/parser/x86_64/x86_64.cpp [hunk 8]
```diff
@@ -158,7 +111,7 @@ namespace smeagle::x86_64 {
         st::Type *paramType = param->getType();
 
         // Get register class based on type
-        std::pair<RegisterClass, RegisterClass> regClasses = getRegisterClassFromType(paramType);
+        Class c = getRegisterClassFromType(paramType);
```
Comment (vsoch) [inline source/parser/x86_64/x86_64.cpp:114]: yeah this line is what threw me of... "A struct called Class?!"

File: test/source/libs/gen_allocation.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+if options.func_filename is None:
+    parser.print_help()
+    exit()
+
+# ------------------------------------------ #
+
+types = [
+    {
+        'category': "Integral Types",
+        'types': [
```
Comment (vsoch) [inline test/source/libs/gen_allocation.py:19]: Ah this could serve as a nice lookup - we could convert this (and other similar tests) to docs when all is said and done. It's nice to see them in one place.

**Comment by vsoch** [comment]
And just curious what is the reason for the Python function over just keeping everything in C++?

File: source/parser/x86_64/allocators.hpp (hunk 1 above)
Comment (hainest) [reply source/parser/x86_64/allocators.hpp:76]: The `getRegistersString` function wasn't correct because x86 doesn't allocate register space for both pieces of the classification. Rather, their characteristics are combined to allocate a single register. For example, an aggregate with two 32-bit `int`s would have the classification `{INTEGER,INTEGER}`, but would only get allocated a single register.

File: source/parser/x86_64/allocators.hpp (hunk 2 above)
Comment (hainest) [reply source/parser/x86_64/allocators.hpp:103]: Ah, yes. That should have been merged together. I also missed the `X86UP` case there. I will fix it!

File: source/parser/x86_64/classifiers.hpp (hunk 3 above)
Comment (hainest) [reply source/parser/x86_64/classifiers.hpp:16]: The rules for classifying aggregates (at least on x86) require looking at only two eightbytes at a time, so there would only ever be at most two.

File: source/parser/x86_64/classifiers.hpp (hunk 4 above)
Comment (hainest) [reply source/parser/x86_64/classifiers.hpp:44]: This is a reminder comment to revisit some architectural designs. As I mentioned in my email last week and Matt brought up in the meeting, I think that we're getting too close to being x86 specific. The ABI model doc also has no discussion of multiple indirections- e.g., pointer to pointer or reference to pointer. I _think_ the current description will compose, but I'm not sure. In general, I think that now would be a good time for us to discuss some more high-level architecting for Smeagle in terms of separating platform-specific information from the generated model facts. I think we're definitely on the right track, overall, but this text about pointers made me really sit down and think about it. If you have some time this week, we should definitely chat!

File: source/parser/x86_64/classifiers.hpp (hunk 5 above)
Comment (hainest) [reply source/parser/x86_64/classifiers.hpp:88]: The are currently placeholders until we settle on something that is abstract enough to encompass all three architectures. We should definitely start a conversation on this in the model doc.

File: source/parser/x86_64/x86_64.cpp (hunk 6 above)
Comment (hainest) [reply source/parser/x86_64/x86_64.cpp:75]: Less is more!

File: source/parser/x86_64/classifiers.hpp (hunk 7 above)
Comment (hainest) [reply source/parser/x86_64/classifiers.hpp:23]: I'm 100% with you on this one. This naming scheme was lifted whole-cloth from the LLVM implementation. Maybe something like `classification` would be a good return type from a function called `classify`?

File: source/parser/x86_64/x86_64.cpp (hunk 8 above)
Comment (hainest) [reply source/parser/x86_64/x86_64.cpp:114]: It's like Java!

File: test/source/libs/gen_allocation.py (hunk 9 above)
Comment (hainest) [reply test/source/libs/gen_allocation.py:19]: Yeah, definitely. I felt it was a good start at trying to separate out what we want to test versus how we implement it. Some testing docs will definitely be in our future!

**Comment by hainest** [comment]
> And just curious what is the reason for the Python function over just keeping everything in C++?

The Python script is there to help generate combinations of test features so we don't have to write hundreds of versions of the functions ourselves. For example, we'll want to add cases for pointers, pointers to pointers, references to pointers, and const versions of all of that. We can just change a few lines of python to generate those hundreds of test cases for us. Eventually, the script will go away once we have formal documentation and we've settled on what we want for test cases.

**Comment by hainest** [comment]
The build failed because it's using Dyninst 11.0.1 instead of HEAD. That's... weird. I tested everything in the container before pushing.

File: source/parser/x86_64/classifiers.hpp (hunk 7 above)
Comment (vsoch) [reply source/parser/x86_64/classifiers.hpp]: yes!

**Comment by vsoch** [comment]
@hainest it could be that the base that is being pulled is not the same as the one you built - if you delete your autamus dyninst container and have it pull from the registry does it still work?

**Comment by vsoch** [comment]
If the one in the registry doesn't work (and this is going to be an issue) we can add building the Dockerfile.dyninst in the PR and then use the latest on Dyninst master. Let me know if the registry one works!

**Comment by hainest** [comment]
I nuked all my docker images. Retesting.

**Comment by vsoch** [comment]
I think the issue was the testing base container wasn't updated - I just made the changes in each test file (lets see if it works!) We probably don't need to build every time (slower testing) given that the updates to the master tag aren't super frequent (I'm happy to do that whenever we need).

**Comment by hainest** [comment]
Ah, ok. Let me pull these updates and I'll rebuild from scratch.

**Comment by vsoch** [comment]
Woohoo! Builds!

**Review by vsoch — APPROVED** [review]
@hainest I'm happy with the changes - are you all set/ready to merge?

**Comment by hainest** [comment]
Merge away!
